sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala

```diff
@@ -160,6 +161,30 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
     this
   }
 
+  /**
+   * Execute a random DDL/DML command inside an external execution engine rather than Spark,
+   * especially for JDBC data source. This could be useful when user has some custom commands
+   * which Spark doesn't support, need to be executed. Please note that this is not appropriate
+   * for query which returns lots of data.
+   *
+   * Datasource should implement <code>ExternalCommandRunnableProvider</code> to perform its
+   * own logic of command execution.
+   *
+   * @since 3.0.0
+   */
+  def executeCommand(command: String): DataFrame = {
+    DataSource.lookupDataSource(source, sparkSession.sessionState.conf) match {
+      case provider if classOf[ExternalCommandRunnableProvider].isAssignableFrom(provider) =>
+        Dataset.ofRows(sparkSession,
+          ExternalCommandExecutor(command, extraOptions.asJava,
+            provider.newInstance().asInstanceOf[ExternalCommandRunnableProvider]))
+
+      case _ =>
+        throw new AnalysisException(s"`executeCommand` is not allowed for source: $source, " +
+          s"because it doesn't implement `ExternalCommandRunnableProvider`")
```

Comment:
I would just say "Command execution is not supported in source $source" because that's what implies by not inheriting `ExternalCommandRunnableProvider`.